nvse/nvse/NVSECompiler.cpp

```diff
@@ -0,0 +1,898 @@
+
+        // OP_LEN
+        AddU16(0x02);
+
+        // Placeholder JMP_OPS
+        auto elsePatch = AddU16(0x0);
+
+        // Compile else block
+        SetU16(elsePatch, CompileBlock(stmt->elseBlock, true));
+    }
+
+    // OP_ENDIF
+    statementCounter.top()++;
+    AddU16(static_cast<uint16_t>(ScriptParsing::ScriptStatementCode::EndIf));
+    AddU16(0x0);
+}
+
+void NVSECompiler::VisitReturnStmt(ReturnStmt* stmt) {
+    // Compile SetFunctionValue if we have a return value
+    if (stmt->expr) {
+        StartCall(0x1546);
+        AddCallArg(stmt->expr);
+        FinishCall();
+    }
+
+    // Emit op_return
+    AddU32(static_cast<uint32_t>(ScriptParsing::ScriptStatementCode::Return));
+}
+
+void NVSECompiler::VisitContinueStmt(ContinueStmt* stmt) {
+    if (loopIncrements.top()) {
+        loopIncrements.top()->Accept(this);
+        statementCounter.top()++;
+    }
+
+    AddU32(0x153E);
+}
+
+void NVSECompiler::VisitBreakStmt(BreakStmt* stmt) {
+    AddU32(0x153F);
+}
+
+void NVSECompiler::VisitWhileStmt(const WhileStmt* stmt) {
+    // OP_WHILE
+    AddU16(0x153B);
+
+    // Placeholder OP_LEN
+    auto exprPatch = AddU16(0x0);
+    auto exprStart = data.size();
+
+    auto jmpPatch = AddU32(0x0);
+
+    // 1 param
+    AddU8(0x1);
+
+    // Compile / patch condition
+    auto condStart = data.size();
+    auto condPatch = AddU16(0x0);
+    insideNvseExpr.push(true);
+    stmt->cond->Accept(this);
+    insideNvseExpr.pop();
+    SetU16(condPatch, data.size() - condStart);
+
+    // Patch OP_LEN
+    SetU16(exprPatch, data.size() - exprStart);
+
+    // Compile block
+    loopIncrements.push(nullptr);
+    CompileBlock(stmt->block, true);
+    loopIncrements.pop();
+    
+    // OP_LOOP
+    AddU16(0x153c);
+    AddU16(0x0);
+    statementCounter.top()++;
+
+    // Patch jmp
+    SetU32(jmpPatch, data.size());
+}
+
+uint32_t NVSECompiler::CompileBlock(StmtPtr stmt, bool incrementCurrent) {
+    for (int i = 0; i < statementCounter.size(); i++) {
+        CompDbg("  ");
+    }
+    CompDbg("Entering block\n");
+    
+    // Get sub-block statement count
+    statementCounter.push(0);
+    stmt->Accept(this);
+    auto newStatementCount = statementCounter.top();
+    statementCounter.pop();
+
+    // Lambdas do not add to parent block stmt count
+    // Others do
+    if (incrementCurrent) {
+        statementCounter.top() += newStatementCount;
+    }
+
+    for (int i = 0; i < statementCounter.size(); i++) {
+        CompDbg("  ");
+    }
+    CompDbg("Exiting Block. Size %d\n", newStatementCount);
+
+    return newStatementCount;
+}
+
+void NVSECompiler::VisitBlockStmt(BlockStmt* stmt) {
+    for (auto& statement : stmt->statements) {
+        statement->Accept(this);
+        statementCounter.top()++;
+    }
+}
+
+void NVSECompiler::VisitAssignmentExpr(AssignmentExpr* expr) {
+    // Assignment as standalone statement
+    if (!insideNvseExpr.top()) {
+        StartCall(0x1539);
```

Comment:
Since these opcodes are not available in GECK command infos it'd be cleaner if an enum were created for them